src/instructlab/model/backends/backends.py

```diff
@@ -208,82 +219,75 @@ def ensure_server(
-    except ClientException:
-        port = free_tcp_ipv4_port(host)
-        logger.debug("Using %i", port)
-
-        host_port = f"{host}:{port}"
-        temp_api_base = get_api_base(host_port)
-        logger.debug(f"Starting a temporary server at {temp_api_base}...")
-        llama_cpp_server_process = None
-        vllm_server_process = None
-
-        if backend == VLLM:
-            # TODO: resolve how the hostname is getting passed around the class and this function
-            vllm_server_process = server_process_func(port=port)
-            count = 0
-            # TODO should this be configurable?
-            vllm_startup_timeout = 300
-            while count < vllm_startup_timeout:
-                sleep(1)
-                try:
-                    list_models(
-                        api_base=temp_api_base,
-                        http_client=http_client,
-                    )
-                    logger.debug(f"model at {temp_api_base} served on vLLM")
-                    break
-                except ClientException:
-                    count += 1
-
-            if count >= vllm_startup_timeout:
-                shutdown_process(vllm_server_process, 20)
-                # pylint: disable=raise-missing-from
-                raise ServerException(
-                    f"vLLM failed to start up in {vllm_startup_timeout} seconds"
-                )
 
-        elif backend == LLAMA_CPP:
-            # server_process_func is a function! we invoke it here and pass the port that was determined
-            # in this ensure_server() function
-            llama_cpp_server_process = server_process_func(port=port)
-            llama_cpp_server_process.start()
-
-            # in case the server takes some time to fail we wait a bit
-            logger.debug("Waiting for the server to start...")
-            count = 0
-            while llama_cpp_server_process.is_alive():
-                sleep(0.1)
-                try:
-                    list_models(
-                        api_base=temp_api_base,
-                        http_client=http_client,
-                    )
-                    break
-                except ClientException:
-                    pass
-                if count > 50:
-                    logger.error("failed to reach the API server")
-                    break
-                count += 1
+    port = free_tcp_ipv4_port(host)
+    logger.debug("Using %i", port)
 
-            logger.debug("Server started.")
+    host_port = f"{host}:{port}"
+    temp_api_base = get_api_base(host_port)
+    logger.debug(f"Starting a temporary server at {temp_api_base}...")
+    llama_cpp_server_process = None
+    vllm_server_process = None
 
-            # if the queue is not empty it means the server failed to start
-            if queue is not None and not queue.empty():
-                # pylint: disable=raise-missing-from
-                raise queue.get()
+    if backend == VLLM:
```

Comment:
Excellent point

Reply:
This is so called _spaghetti code_: unstructured mix of code with redundant functions and conditionals.

Exactly, this patch is preparation to:
1. move `vllm` and `llama_cpp` code blocks to appropriate files
2. disband `ensure_server()`

Reply:
> This is so called _spaghetti code_: unstructured mix of code with redundant functions and conditionals.

@makelinux Rather than pointing out issues after the fact, I'd suggest you start doing reviews so these issues can be fixed during the code review process instead of sending follow-up PRs. I'm sure your input will be valuable. Thanks!

Reply:
> Apologies if this is obvious, and I know this code was already in this file, but why isn't this logic in vllm and llama_cpp rather than have the large if blocks here?

I've addressed this issues here: #1611